### ISSUE TITLE
Consider providing a pytest plugin

### ISSUE THREAD

**Issue by RobertCraigie** [theacodes/nox#584]
**How would this feature be useful?**

I am currently using a custom pytest setup to run integration tests for my project. This finds any `test.sh` scripts and registers them with pytest, this is not ideal to say the least and I am also managing dependencies myself which causes some complexities, I am planning on switching to use nox directly within these scripts but it would be even better if there could be some way to define these tests and let nox handle everything for me.

An example of a test script can be found here: https://github.com/RobertCraigie/prisma-client-py/blob/main/tests/integrations/postgresql/test.sh and the plugin that finds and registers the script with pytest can be found here: https://github.com/RobertCraigie/prisma-client-py/blob/main/tests/integrations/conftest.py

**Describe the solution you'd like**

I'm not sure what the API for this would look like, just throwing the idea out there!

**Describe alternatives you've considered**

I am in the process of switching the custom test scripts to use nox to avoid handling dependencies myself. Once this is done, a plugin provided directly by nox would just be a QOL improvement and would allow me to remove some code.

I appreciate that this could very much be out of scope for nox to provide.

**Comment by cjolowicz** [comment]
I'm not sure I understand why you are using pytest plugins.

From a quick look at your code, I would advise the following:

Place a noxfile.py at the top level of your repository (or your integration test suite). Define two Nox sessions, one for testing, one for type-checking.

The test session has two parts:

1. The session installs the test requirements and your package(s). Merge your test requirements files into a single one.
2. The session invokes pytest (via coverage). Let pytest discover the test modules.

Convert the various test setups (e.g. `prisma db push`) into session-scoped pytest fixtures. Ensure that these fixtures are only used for the tests that need them.

The type-checking session installs pyright and runs it.

Does that make sense?

**Comment by RobertCraigie** [comment]
> I'm not sure I understand why you are using pytest plugins.

I use pytest plugins and separate environments for the integration tests as they are testing auto-generated library code, each environment will have different code for the same library meaning they cannot be shared in the same environment. 

However, rethinking this, I may be able to merge them into a single environment by auto-generating code to different locations, this should work but would make it awkward to use as I would then have to import the different integration tests like so:

```py
# instead of
import prisma
# it would have to be something like this
import prisma_postgres
```

I'm pretty sure that this solution would also mean that I wouldn't be able make use of a separate pytest plugin I plan on adding (https://github.com/RobertCraigie/prisma-client-py/issues/171) or share any code between the tests.

Another downside with this is the loss of separation of concerns:

- the feedback loop would be longer when running individual tests as the clients for every test would have to be generated instead of the one client needed for the given test.
- I don't want contributors to have to install / setup database servers for every supported database provider

**Comment by RobertCraigie** [comment]
> However, I still don't understand why you are using pytest plugins. You say the pytest plugin locates test.sh scripts, and I can see that those scripts, in turn, invoke pytest. So pytest calling pytest? If you could clear this up for me, we may be able to figure out how Nox might help simplify these workflows.

I decided to use a pytest plugin so that the `test.sh` would be picked up by pytest so that these tests are included in the main workflow. It should be noted that not all `test.sh` scripts call pytest.

> If I understand your feature request correctly, you want pytest to be able to call Nox as a plugin.

Essentially what I would like from this feature is the ability to write nox sessions and have pytest pick them up and run them as a standard test. Heres what the noxfile would look like for the postgresql test:

```py
import nox

@nox.session
def postgres(session: nox.Session) -> None:
    session.install('-r', 'requirements.txt')
    session.install('.')

    session.run('prisma', 'db', 'push', '--accept-data-loss', '--force-reset')
    session.run('coverage', 'run', '-m', 'pytest', '--confcutdir=.', 'tests')
    session.run('pyright')
```

I'm thinking that there are two possible options here:

1. Nox provides a pytest fixture that provides a `Session` object
2. Nox could somehow pick up sessions and register them during the pytest collection process

Providing a solution like this would make it very easy to write automated tests for different environments.

> Nox is designed to create environments and install and runs tools like pytest inside them. Your idea inverts this control flow.

Yes, this feature request is only useful for people who would want to run tests within different environments.

> Nox knows nothing about pytest; it's a generic automation framework that just happens to be used with pytest a lot.

Nothing would change for any person using Nox in the standard way it was designed for, pytest plugins are registered through package entry points and are automatically picked up when pytest is ran.

```py
setup(
    ...
    entry_points={
        'pytest11': [
            'nox = nox._pytest',
        ],
    },
)
```

Then the plugin code would go in `nox/_pytest.py` file which would only be imported by pytest.

**Comment by cjolowicz** [comment]
What does the outer pytest add compared to just using plain Nox as the main entry point? 

Drop a single noxfile.py into your repository, with one session per auto-generated library. In other words, each session corresponds to a test.sh script. Running `nox` will create separate environments for each library, and run whatever commands you need to install and test it.

**Comment by RobertCraigie** [comment]
Admittedly it does not add much value, I could refactor it so that the outer pytest call is not required. Explaining my original reasoning did make me realise that there are better methods to achieve what I want.

Thanks for your help!

**Comment by cjolowicz** [comment]
You're welcome!
